cloudinit/config/cc_mounts.py

```diff
@@ -223,6 +223,63 @@ def suggested_swapsize(memsize=None, maxsize=None, fsys=None):
     return size
 
 
+def get_fstype(path):
+    """Return the type of the filesystem that path is on.
+    If path is a mount point itself, then the type of the filesystem mounted
+    on path rather than its parent dir."""
+    return util.get_mount_info(path)[1]
+
+
+def create_swapfile(fname, size):
+    """Size is in MB."""
+
+    def unlink_f(fname):
+        if os.path.exists(fname):
+            os.unlink(fname)
+
+    def create_swap(fname, size, method):
+        errmsg = "Failed to create swapfile '%s' of size %dMB via %s: %s"
+
+        LOG.debug("Creating swapfile in '%s' on fstype '%s' using '%s'",
+                  fname, fstype, method)
+
+        if method == "fallocate":
+            cmd = ['fallocate', '-l', '%dM' % size, fname]
+        elif method == "dd":
+            cmd = ['dd', 'if=/dev/zero', 'of=%s' % fname, 'bs=1M',
+                   'count=%d' % size]
+
+        try:
+            util.subp(cmd, capture=True)
+        except util.ProcessExecutionError as e:
+            LOG.warning(errmsg, fname, size, method, e)
+
+            if method == "fallocate":
+                LOG.warning("Will attempt with dd.")
+                create_swap(fname, size, "dd")
```

Comment:
This behaviour seems a little unexpected to me. I wouldn't expect `create_swap(..., "fallocate")` to do anything other than attempt to use `fallocate` to create the swap.

I think moving this logic up to the calling function would make it clearer what the high-level behaviour is:

```py
    if fstype in ("xfs", "btrfs"):
        create_swap(fname, size, "dd")
    else:
        try:
            create_swap(fname, size, "fallocate")
        except util.ProcessExecutionError as e:
            create_swap(fname, size, "dd")
```

Reply:
Agreed